code/modules/mining/machine_silo.dm

```diff
@@ -194,9 +307,185 @@
+		user.flash_act(intensity = 1, affect_silicon = TRUE)
+		handle_access_action_feedback(
+			BAN_ATTEMPT_FAILURE_SOULLESS_MACHINE,
+			ID_DATA(user),
+			target_user_data
+		)
+		return
+
+	var/target_bank_id = target_user_data["Account ID"]
+	var/target_is_banned = banned_users.Find(target_bank_id)
+	// Agent card bypasses bans but we want to specially handle them anyway
+	// so a bunch of random account IDs don't fill the list and do something
+	// like ban people who haven't joined the round yet
+	var/haxxor_card_ban_immunity = !isnull(target_user_data[CHAMELEON_OVERRIDE])
+
+	to_chat(user, span_warning("You press the button to [target_is_banned ? "un" : ""]ban [target_user_data["Name"]]'s account..."))
+	// No feedback if emagged
+	if(emagged)
+		if(!haxxor_card_ban_immunity && isnum(target_bank_id))
+			target_is_banned ? banned_users.Remove(target_bank_id) : banned_users.Add(target_bank_id)
+		return
+
+	var/alist/silo_user_data = ID_DATA(user)
+	var/list/silo_user_accesses = astype(silo_user_data["Accesses"], /list)
+	var/list/target_user_accesses = astype(target_user_data["Accesses"], /list)
+
+
+	// Even though QM bypasses the access check (or rather always pases)
+	// perhaps the Captain would pre-emptively ban them right before a demotion
+	if(target_user_accesses?.Find(ACCESS_QM) && !silo_user_accesses?.Find(ACCESS_CAPTAIN))
+		handle_access_action_feedback(
+			BAN_ATTEMPT_FAILURE_CHALLENGING_DA_CHIEF,
+			silo_user_data,
+			target_user_data
+		)
+		return
+	if(!silo_user_accesses?.Find(ACCESS_QM))
+		handle_access_action_feedback(
+			BAN_ATTEMPT_FAILURE_NO_ACCESS,
+			silo_user_data,
+			target_user_data
+		)
+		return
+	if(haxxor_card_ban_immunity)
+		handle_access_action_feedback(
+			FAILED_OPERATION_SUSPICIOUS,
+			silo_user_data,
+			target_user_data
+		)
+		return
+	if(!target_bank_id || !isnum(target_bank_id))
+		handle_access_action_feedback(
+			FAILED_OPERATION_NO_BANK_ID,
+			silo_user_data,
+			target_user_data
+		)
+		return
+	if(target_is_banned)
+		banned_users.Remove(target_bank_id)
+		handle_access_action_feedback(
+			UNBAN_CONFIRMATION,
+			silo_user_data,
+			target_user_data
+		)
+		return
+	// If we got here, we are banning the user
+	banned_users.Add(target_bank_id)
+	handle_access_action_feedback(
+		BAN_CONFIRMATION,
+		silo_user_data,
+		target_user_data
+	)
+
+/obj/machinery/ore_silo/proc/attempt_toggle_restrict(mob/living/user)
+	if(!istype(user))
+		CRASH()
+	var/emagged = obj_flags & EMAGGED
+	if(emagged)
+		ID_required = !ID_required
+		return
+	var/alist/silo_user_data = ID_DATA(user)
+	var/is_a_robot = silo_user_data[SILICON_OVERRIDE]
+	if(is_a_robot)
+		handle_access_action_feedback(
+			UNRESTRICT_FAILURE_SOULLESS_MACHINE,
+			silo_user_data,
+			null
+		)
+		return
+	var/list/user_accesses = astype(silo_user_data["Accesses"], /list)
+	if(!user_accesses.Find(ACCESS_QM))
```

Comment:
Runtimes with null.Find is this toggle is called while not wearing a visible ID. Look into an early return potentially for worn IDs